expression/builtin_time.go

```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
 	baseFunctionClass
 }
 
+func (c *timestampFunctionClass) getDefaultFsp4Type(tp *types.FieldType) int {
+	if tp.Tp == mysql.TypeDatetime || tp.Tp == mysql.TypeDate || tp.Tp == mysql.TypeDuration || tp.Tp == mysql.TypeTimestamp || tp.Tp == mysql.TypeNewDate {
+		return tp.Decimal
+	}
+	switch cls := tp.ToClass(); cls {
+	case types.ClassInt:
+        return 0
+	case types.ClassReal, types.ClassString:
+		return types.MaxFsp
+	case types.ClassDecimal:
+        if tp.Decimal < types.MaxFsp {
+			return tp.Decimal
+		}
+		return 6
+	}
+    return types.MaxFsp
+}
+
 func (c *timestampFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinTimestampSig{newBaseBuiltinFunc(args, ctx)}
+	evalTps, argLen := []evalTp{tpString}, len(args)
+	if argLen == 2 {
+		evalTps = append(evalTps, tpString)
+	}
+	fsp := c.getDefaultFsp4Type(args[0].GetType())
+    if argLen == 2 {
+		fsp = mathutil.Max(fsp, c.getDefaultFsp4Type(args[1].GetType()))
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpDatetime, evalTps...)
+    bf.tp.Decimal, bf.tp.Flen = fsp, 19
+	if fsp != 0 {
+		bf.tp.Flen += 1 + fsp
+	}
+
+	if err != nil {
```

Comment:
check `err != nil` first before you use `bf`

Reply:
done